File: api/src/main/java/com/netflix/iceberg/Files.java
/*
 * Licensed to the Apache Software Foundation (ASF) under one
 * or more contributor license agreements.  See the NOTICE file
 * distributed with this work for additional information
 * regarding copyright ownership.  The ASF licenses this file
 * to you under the Apache License, Version 2.0 (the
 * "License"); you may not use this file except in compliance
 * with the License.  You may obtain a copy of the License at
 *
 *   http://www.apache.org/licenses/LICENSE-2.0
 *
 * Unless required by applicable law or agreed to in writing,
 * software distributed under the License is distributed on an
 * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
 * KIND, either express or implied.  See the License for the
 * specific language governing permissions and limitations
 * under the License.
 */

package com.netflix.iceberg;

import com.netflix.iceberg.exceptions.AlreadyExistsException;
import com.netflix.iceberg.exceptions.RuntimeIOException;
import com.netflix.iceberg.io.InputFile;
import com.netflix.iceberg.io.OutputFile;
import com.netflix.iceberg.io.PositionOutputStream;
import com.netflix.iceberg.io.SeekableInputStream;
import java.io.File;
import java.io.FileNotFoundException;
import java.io.IOException;
import java.io.RandomAccessFile;

public class Files {

  public static OutputFile localOutput(File file) {
    return new LocalOutputFile(file);
  }

  public static OutputFile localOutput(String file) {
    return localOutput(new File(file));
  }

  private static class LocalOutputFile implements OutputFile {
    private final File file;

    private LocalOutputFile(File file) {
      this.file = file;
    }

    @Override
    public PositionOutputStream create() {
      if (file.exists()) {
        throw new AlreadyExistsException("File already exists: %s", file);
      }

      try {
        return new PositionFileOutputStream(new RandomAccessFile(file, "rw"));
      } catch (FileNotFoundException e) {
        throw new RuntimeIOException(e, "Failed to create file: %s", file);
      }
    }

    @Override
    public PositionOutputStream createOrOverwrite() {
      if (file.exists()) {
        if (!file.delete()) {
          throw new RuntimeIOException("Failed to delete: " + file);
        }
      }
      return create();
    }

    @Override
    public String location() {
      return file.toString();
    }

    @Override
    public InputFile toInputFile() {
      return localInput(file);
    }

    @Override
    public String toString() {
      return location();
    }
  }

  public static InputFile localInput(File file) {
    return new LocalInputFile(file);
  }

  public static InputFile localInput(String file) {
    return localInput(new File(file));
  }

  private static class LocalInputFile implements InputFile {
    private final File file;

    private LocalInputFile(File file) {
      this.file = file;
    }

    @Override
    public long getLength() {
      return file.length();
    }

    @Override
    public SeekableInputStream newStream() {
      try {
        return new SeekableFileInputStream(new RandomAccessFile(file, "r"));
      } catch (FileNotFoundException e) {
        throw new RuntimeIOException(e, "Failed to read file: %s", file);
      }
    }

    @Override
    public String location() {
      return file.toString();
    }

    @Override
    public String toString() {
      return location();
    }
  }

  private static class SeekableFileInputStream extends SeekableInputStream {
    private final RandomAccessFile stream;

    private SeekableFileInputStream(RandomAccessFile stream) {
      this.stream = stream;
    }

    @Override
    public long getPos() throws IOException {
      return stream.getFilePointer();
    }

    @Override
    public void seek(long newPos) throws IOException {
      stream.seek(newPos);
    }

    @Override
    public int read() throws IOException {
      return stream.read();
    }

    @Override
    public int read(byte[] b) throws IOException {
      return stream.read(b);
    }

    @Override
    public int read(byte[] b, int off, int len) throws IOException {
      return stream.read(b, off, len);
    }

    @Override
    public long skip(long n) throws IOException {
      if (n > Integer.MAX_VALUE) {
        return stream.skipBytes(Integer.MAX_VALUE);
      } else {
        return stream.skipBytes((int) n);
      }
    }

    @Override
    public void close() throws IOException {
      stream.close();
    }
  }

  private static class PositionFileOutputStream extends PositionOutputStream {
    private final RandomAccessFile stream;

    private PositionFileOutputStream(RandomAccessFile stream) {
      this.stream = stream;
    }

    @Override
    public long getPos() throws IOException {
      return stream.getFilePointer();
    }

    @Override
    public void write(byte[] b) throws IOException {
      stream.write(b);
    }

    @Override
    public void write(byte[] b, int off, int len) throws IOException {
      stream.write(b, off, len);
    }

    @Override
    public void close() throws IOException {
      stream.close();
    }

    @Override
    public void write(int b) throws IOException {
      stream.write(b);
    }
  }
}

File: python/iceberg/api/files.py
# Licensed to the Apache Software Foundation (ASF) under one
# or more contributor license agreements.  See the NOTICE file
# distributed with this work for additional information
# regarding copyright ownership.  The ASF licenses this file
# to you under the Apache License, Version 2.0 (the
# "License"); you may not use this file except in compliance
# with the License.  You may obtain a copy of the License at
#
#   http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing,
# software distributed under the License is distributed on an
# "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY
# KIND, either express or implied.  See the License for the
# specific language governing permissions and limitations
# under the License.

import gzip
import os

from iceberg.exceptions import AlreadyExistsException

from .expressions import JAVA_MAX_INT
from .io import (InputFile,
                 OutputFile,
                 PositionOutputStream,
                 SeekableInputStream)


class Files(object):

    @staticmethod
    def local_output(file):
        return LocalOutputFile(file)

    @staticmethod
    def local_input(file):
        return LocalInputFile(file)


class LocalOutputFile(OutputFile):

    def __init__(self, file):
        self.file = file

    def create(self):
        if os.path.exists(self.file):
            raise AlreadyExistsException("File already exists: %s" % self.file)

        return PositionOutputStream(open(self.file, "w"))

    def create_or_overwrite(self):
        if os.path.exists(self.file):
            if not os.remove(self.file.name):
                raise RuntimeError("File deleting file: %s" % self.file)

        return self.create()

    def location(self):
        return self.file

    def __str__(self):
        return self.location()


class LocalInputFile(InputFile):

    def __init__(self, file):
        self.file = file

    def get_length(self):
        return os.path.getsize(self.file)

    def new_stream(self, gzipped=False):
        with open(self.file, "rb") as fo:
            if gzipped:
                fo = gzip.GzipFile(fileobj=fo)
            return fo

    def location(self):
        return self.file

    def __str__(self):
        return self.location()


class SeekableFileInputStream(SeekableInputStream):

    def __init__(self, stream):
        self.stream = stream

    def get_pos(self):
        return self.stream.tell()

    def seek(self, new_pos):
        return self.stream.seek(new_pos)

    def read(self, b=None, off=None, read_len=None):
        if b is None and off is None and read_len is None:
            return None, self.stream.read()
        if b is None:
            raise RuntimeError("supplied byte field is None")

        if off is None and read_len is None:
            new_b = self.stream.read(b.length)
            return len(new_b), new_b

        if off is not None and read_len is None or off is None and read_len is not None:
            raise RuntimeError("Invalid args: read_len(%s), off(%s)" % (read_len, off))

        if read_len < 0 or off < 0 or (len(b) - off < read_len):
            raise RuntimeError("Invalid args: read_len(%s), off(%s), len_b_offset(%s)" % (read_len, off, len(b) - off))

        new_b = bytes(self.stream.read(read_len), "utf8")

        if off > 0:
            new_b = b[0:off] + new_b
        if off + read_len < len(b):
            new_b = new_b + b[off + read_len:]
        return read_len, new_b

    def skip(self, n):
        if n > JAVA_MAX_INT:
            return self.stream.seek(JAVA_MAX_INT)
        else:
            return self.stream.seek(n)
